### PR TITLE
Restructure modules to make spatial clustering algorithms reusable

### LINKED ISSUE #492: Restructure cluster module
The cluster module is very clustered currently. Since it should cluster electrical line or gas pipeline networks and not itself, the code should be rewritten to be more clear. The following structure is proposed to make maintainance more easy and allow future development to reuse existing parts of the code:

- sector specific preprocessing (electrical.py, gas.py)
- spatial clustering (spatial.py)
- sectorspecific postprocessing (electrical.py, gas.py)

By this structure, the sectors can be individually prepared for the clustering, then sharing the same methods by providing data in a uniform datastructure and getting back the results in the same uniform datastructre to follow up in a sector specific postprocessing.

### PR DESCRIPTION
**Pull request by fwitte**
Resolve #492 

To make use of the implemented algorithm, I splitted the methods defined in the former "networkclustering.py", which was separated into etrago/cluster/spatial and etrago/cluster/electrical. etrago/cluster/gas will make use of the methods in etrago/cluster/spatial.

Unfortunately, I removed history for the electrical.py, however git blame still works. If that is an issue, I can use git-filter-repo to keep history for both files (see https://stackoverflow.com/a/61917589/12690690). I worked with it in different projects, it is a really nice tool.

Also, it seems that some preprocessing and postprocessing steps are identical for the different algorithms, so we should make use of that fact and reduce lines of code.

### CODE REVIEW

**Comment by fwitte** [comment]
@pieterhexen and @ClaraBuettner: The name of this PR + branch is currently not really correct, please ignore that.

I made a first draft to structure the clustering methods:

- sector specific preprocessing
- clustering
- sectorspecific postprocessing

I tested the implementation, it produces identical results for dev branch and this branch for both electrical and gas system. The kmedoids clustering for the gas system is not yet implemented here, but we might want to move this to dev already, because it makes a lot of things much clearer, I think.

What is your opinion about it?

Thank you! :)

**Review by pieterhexen — CHANGES_REQUESTED** [review]
Checked code and tested on test_dump. Everything works fine.

File: etrago/appl.py [hunk 1]
```diff
@@ -92,11 +92,13 @@
     "extra_functionality": {},  # Choose function name or {}
     # Clustering:
     "network_clustering": {
+        "random_state": 123456,  # random state for replicability of kmeans results
```
[the remark this answers is not in the document]
Comment (fwitte) [reply etrago/appl.py]: Critical issue resolved by [ee0bddd](https://github.com/openego/eTraGo/pull/485/commits/ee0bdddeee537d28b16c08eedee03fb1a3a1458f), thank you!

**Comment by CarlosEpia** [comment]
Hi @fwitte. I see big changes in this PR, therefore it would be nice if you could merge dev into this branch and check if it works correctly together with the last changes in dev. Additionally, can you describe briefly the changes you did? Since the history of electrical.py is not available, that description would help me to choose what parts of the code review more carefully.

**Comment by CarlosEpia** [comment]
I just tried to run this branch using the eHV clustering and ran into this error message:

  File "/home/student/git/eTraGo/etrago/appl.py", line 513, in <module>
    etrago = run_etrago(args, json_path=None)

  File "/home/student/git/eTraGo/etrago/appl.py", line 475, in run_etrago
    etrago.ehv_clustering()

  File "/home/student/git/eTraGo/etrago/cluster/electrical.py", line 361, in ehv_clustering
    busmap = busmap_from_psql(self)

NameError: name 'busmap_from_psql' is not defined

**Comment by fwitte** [comment]
@CarlosEpia: Are you available today?

**Comment by CarlosEpia** [comment]
> @CarlosEpia: Are you available today?

Yeah, sure. If it is ok for you, we can meet in the Jitsi room at 11:00.

**Comment by CarlosEpia** [comment]
@fwitte according to our meeting, I will check this PR as soon as possible. Since the purpose of this branch is different from its original issue, could you create a new issue to describe the problem that this PR is solving? It is just for traceability.

**Comment by fwitte** [comment]
Okay, everything ready here, @pieterhexen and @CarlosEpia, there is also a respective issue #492

File: etrago/appl.py [hunk 2]
```diff
@@ -92,11 +92,13 @@
     "extra_functionality": {},  # Choose function name or {}
     # Clustering:
     "network_clustering": {
+        "random_state": 42,  # random state for replicability of kmeans results
         "active": True,  # choose if clustering is activated
         "method": "kmedoids-dijkstra",  # choose clustering method: kmeans or kmedoids-dijkstra
         "n_clusters_AC": 30,  # total number of resulting AC nodes (DE+foreign)
         "cluster_foreign_AC": False,  # take foreign AC buses into account, True or False
-        "n_clusters_gas": 30,  # total number of resulting CH4 nodes (DE+foreign)
+        "method_gas": "kmeans",  # choose clustering method: kmeans or kmedoids-dijkstra
```
Comment (CarlosEpia) [inline etrago/appl.py:100]: Please update the description of the parameters that were changed in this PR

**Review by CarlosEpia — APPROVED** [review]
This branch is apparently working as expected. I ran eHV and Dijktra clustering successfully using the test case and full dataset. It can be merged after updating the parameters' description in appl.py

**Comment by fwitte** [comment]
> apparently

:D

> It can be merged after updating the parameters' description in appl.py

Updated the description, you can merge it @pieterhexen or @CarlosEpia

**Comment by CarlosEpia** [comment]
> > apparently
> 
> :D
> 
> > It can be merged after updating the parameters' description in appl.py
> 
> Updated the description, you can merge it @pieterhexen or @CarlosEpia

Hi @pieterhexen does it make sense for you if you merge this branch since I can see that #493  is also approved?

**Comment by pieterhexen** [comment]
> Hi @pieterhexen does it make sense for you if you merge this branch since I can see that #493 is also approved?

Hey Carlos, I'll merge the change in #493 later because it is not a bug and this branch doesn't change the behaviour of the clustering.

**Pull request merged**